### PR TITLE
[ Don't merge!!! ] Add Test-Json cmdlet

### LINKED ISSUE #4220: Test-Json cmdlet (JSON lint)
Placeholder for now as I follow up on some customer scenarios: it would be really nice to have a cmdlet that could validate JSON before it's used as input. I'd expect for something like this to already exist in Newtonsoft.Json. 

IMHO, it's a nice-to-have for 6.0.0 depending on how we're looking with `HighPriority` over the next couple months. If not, we could move it out to 6.1.0 without a real problem.

### PR DESCRIPTION
**Pull request by iSazonov**
Resolve #4220.

The cmdlet is based on [Newtonsoft.Json.Schema](https://github.com/JamesNK/Newtonsoft.Json.Schema).
It allow to check:
- Json by only parsing
- Json against Schema
- implicitly check Schema by parsing (based on previous line check)

Please see [license](https://github.com/JamesNK/Newtonsoft.Json.Schema/blob/master/LICENSE.md) of the framework (GNU Affero General Public License). /cc @joeyaiello @SteveL-MSFT

### CODE REVIEW

**Comment by SteveL-MSFT** [comment]
I’ll check if we can accept this license.  For now, you’ll need to add it to https://github.com/PowerShell/PowerShell/blob/master/ThirdPartyNotices.txt as part of this PR

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (markekraus) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Can this include a note that it implements the `Test-Json` command?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (markekraus) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Add newline

**Comment by iSazonov** [comment]
@SteveL-MSFT I can not pull changes in ThirdPartyNotices.txt - it is excluded file.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Done.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Fixed.

**Comment by SteveL-MSFT** [comment]
@iSazonov ok, I'll take care of that file then once I get approval.  We should not merge this PR until the license is approved for this project.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: I don't know what the process is for creating Cmdlet help/helpuri.  Should we have an issue tracking this so that the work item is not forgotten?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: I think [ValidateNotNullOrEmpty()] is needed here too.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: I don't think this check is necessary since Schema property is validated as not null.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: [Nitpick] We normally don't explicitly initialize fields to null since C# guarantees fields are initialized to zero.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: I don't think this null check is necessary since we throw terminating error for invalid schema.

**Review by PaulHigin — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: @Francisco-Gamino @daxian-dbw Could you help with new HelpUri?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: @daxian-dbw It seems we should add this in our code guidance.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: My thoughts is:
 - Here we can get values from pipeline
 - The cmdlet is specific - it _tests_ input objects 
-- so for consistency it is better to be based on Json library check logic.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: It is a check that the property present.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1]: First 4 It can be converted to one with testcases and splatting.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1
Comment (adityapatwardhan) [inline test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1]: This one and the one above can be combined into single It with testcases.

**Review by adityapatwardhan — CHANGES_REQUESTED** [review]

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Above we check whether Schema present - it is optional parameter. If Schema is absent here _jschema is null and we skip the test aganist a json schema.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1]: Done.

File: test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1
Comment (iSazonov) [reply test/powershell/Modules/Microsoft.PowerShell.Utility/Test-Json.Tests.ps1]: I tried to do it, but there are too many parameters, and these tests are losing visibility. Maybe leave as is?

**Comment by SteveL-MSFT** [comment]
To set expectation, since this depends on a component that has a non-permissive license, it may take some time to get this approved.  Also, since the [Ignite](https://www.microsoft.com/en-us/ignite/default.aspx) conference is this week, many of the approvers are not here.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Ok.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: This is a required property.  Json check cannot be run without a valid schema property.  Make this property "mandatory" and let PowerShell do the validation rather than have an explicit check.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Schema should be a required "mandatory" property since the Json check cannot be run without it.  Make Schema mandatory and then the two explicit null checks are not necessary.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: What if an user want only check json parsing? Why we should block the scenario?

Comment (PaulHigin) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Ok, I understand now.  Can you add a comment making it clear that Schema is optional and without it Json parsing is still performed and of value?  I assume this validates that the json is well formed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [inline src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: I am not familiar with this framework.  Do we need to be concerned about code injection here when passing the input string directly to the Parse() API?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: I added the comment in Schema parameter description and in the PR description.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: The initializer was removed.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Good question.
@JamesNK Could you please comment? Should we be concerned about code injection in `JObject.Parse(Json)` ?

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (PaulHigin) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: Please disregard.

File: src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs
Comment (iSazonov) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: We can look at the [Parse source code](https://github.com/JamesNK/Newtonsoft.Json/blob/master/Src/Newtonsoft.Json/Linq/JObject.cs#L423). It uses `StringReader` at low level and then tokenize the string stream. 
I don't see threats but I haven't an experience in the security area.

Comment (PaulHigin) [reply src/Microsoft.PowerShell.Commands.Utility/commands/utility/TestJsonCommand.cs]: I agree that it is an unlikely threat with this Json reader/parser, but I wanted to raise awareness for possible code injection vulnerabilities when we pass external input to a 3rd party API.

**Comment by TravisEz13** [comment]
FYI, there are merge conflicts.

**Comment by iSazonov** [comment]
Rebased to resolve merge conflicts.

**Comment by iSazonov** [comment]
@SteveL-MSFT Could you ping license approvers? Also can we ask the author of the package to change/add MIT license?

**Comment by SteveL-MSFT** [comment]
@iSazonov since it appears that [Json.Net Schema](https://www.newtonsoft.com/jsonschema) is how NewtonSoft makes money, I doubt they will change the license.  Still waiting on the approval, but to be honest with a non-permissive license, this will likely not be approved as I don't think it's compatible with the MIT license.  Your best option might be to publish this as a separate module on the PowerShellGallery as long as it adheres to the AGPL.

**Comment by iSazonov** [comment]
@SteveL-MSFT Thanks. I hope that if they offer AGPL for open source projects, they don't lose anything with adding MIT.

**Comment by SteveL-MSFT** [comment]
@iSazonov so after talking to legal, the end result is that if we accept this PR which will pull in the AGPL assembly and we distribute that package, then we would have to re-license PSCore6 under the AGPL to comply with the AGPL.  I think the alternate path is for you to publish this code under AGPL in your own repo and publish the resulting module on PowerShellGallery and then there is no issue with someone using a AGPL module with our MIT PSCore6.

**Comment by iSazonov** [comment]
@SteveL-MSFT I've made a request to obtain a suitable license.
If the answer will be no, I guess we have to find another way, because Json is very desirable to have inside.

**Comment by iSazonov** [comment]
Because of the licensing problem, I closed the PR and I'll open a new PR with another library.

**Pull request closed without merging**
